components/encryption/src/master_key/kms.rs

```diff
@@ -148,92 +197,71 @@ where
     Err(Error::Other(box_err!("{:?}", last_err)))
 }
 
-struct Inner {
-    config: KmsConfig,
-    backend: Option<MemAesGcmBackend>,
-    cached_ciphertext_key: Vec<u8>,
+#[derive(Debug)]
+struct State {
+    encryption_backend: MemAesGcmBackend,
+    cached_ciphertext_key: Encrypted,
 }
 
-impl Inner {
-    fn maybe_update_backend(&mut self, ciphertext_key: Option<&Vec<u8>>) -> Result<()> {
-        let http_dispatcher = HttpClient::new().unwrap();
-        self.maybe_update_backend_with(ciphertext_key, http_dispatcher)
+impl State {
+    fn new_from_datakey(datakey: DataKey) -> Result<State> {
+        Ok(State {
+            cached_ciphertext_key: datakey.encrypted,
+            encryption_backend: MemAesGcmBackend::new(datakey.plaintext)?,
+        })
     }
 
-    fn maybe_update_backend_with<D>(
-        &mut self,
-        ciphertext_key: Option<&Vec<u8>>,
-        dispatcher: D,
-    ) -> Result<()>
-    where
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        if self.backend.is_some()
-            && ciphertext_key.map_or(true, |key| *key == self.cached_ciphertext_key)
-        {
-            return Ok(());
-        }
-
-        let mut kms = AwsKms::with_request_dispatcher(&self.config, dispatcher)?;
-        let key = if let Some(ciphertext_key) = ciphertext_key {
-            self.cached_ciphertext_key = ciphertext_key.to_owned();
-            kms.decrypt(ciphertext_key)?
-        } else {
-            let (ciphertext_key, plaintext_key) = kms.generate_data_key()?;
-            self.cached_ciphertext_key = ciphertext_key;
-            plaintext_key
-        };
-        if self.cached_ciphertext_key == key {
-            panic!(
-                "ciphertext key should not be the same as master key, \
-                otherwise it leaks master key!"
-            );
-        }
-
-        // Always use AES 256 for encrypting master key.
-        self.backend = Some(MemAesGcmBackend::new(key)?);
+    fn update_from_datakey(&mut self, datakey: DataKey) -> Result<()> {
+        self.cached_ciphertext_key = datakey.encrypted;
+        // Always use AES 256
+        self.encryption_backend = MemAesGcmBackend::new(datakey.plaintext)?;
         Ok(())
     }
+
+    fn cached(&self, ciphertext_key: &Encrypted) -> bool {
+        *ciphertext_key == self.cached_ciphertext_key
+    }
+}
+
+pub struct DataKey {
+    pub encrypted: Encrypted,
+    pub plaintext: Vec<u8>,
 }
 
+pub trait KmsProvider: Sync + Send + 'static + std::fmt::Debug {
+    fn generate_data_key(&mut self) -> Result<DataKey>;
+    fn decrypt(&mut self, ciphertext: EncryptedRef) -> Result<Vec<u8>>;
+    fn name(&self) -> &[u8];
+}
+
+#[derive(Debug)]
 pub struct KmsBackend {
-    inner: Mutex<Inner>,
+    state: Mutex<State>,
+    kms_provider: Mutex<Box<dyn KmsProvider>>,
```

Comment:
Since `KmsProvider` is `Sync`, we don't need a mutex to guard it.

Reply:
The Mutex is to avoid requiring an `&mut` for `KmsBackend` in the `decrypt` function.

Reply:
why you want to avoid `&mut`?

Reply:
The other backends and the interface don't need `&mut`, and `encrypt` does not need it, so I was hesitant to change the interface to `&mut` just for `decrypt`: to me that interface definition would look odd. I suppose I can add some comments saying that it is mutable for the purposes of caching internal state.

Reply:
I think you can remove `&mut` from `KmsProvider::generate_data_key`, then you don't need a mutex for the `kms_provider`.

Reply:
In `generate_data_key` `retry()` requires `runtime` to be mutable.

Reply:
Not familiar with tokio, but maybe `runtime` should not be member of `AwsKms`, instead being created for each invocation of `decrypt`?

Reply:
cc @overvenus

Reply:
Runtimes are generally designed to be persistent. I suspect we should not be creating a runtime but instead passing one in from a top level. That will still require a mutable runtime. However, in either case if we assume all KmsProvider needs the tokio runtime we can pass it in to the decrypt_data_key function.

Reply:
Okay, runtime is now in State and passed into the backend, so the backend is now immutable and does not need a Mutex.

Reply:
cool